pulser/simulation/simresults.py

```diff
@@ -14,63 +14,305 @@
+            # ground-rydberg basis case
+            else:
+                proj = qutip.tensor([qutip.basis(2, 1-int(i)).proj() for i
+                                     in bitstring])
+            return proj
+
+        return sum(v * _proj_from_bitstring(b) for
+                   b, v in self._results[t].items())
+
+    def get_final_state(self) -> qutip.Qobj:
+        """Get the final state of the simulation as a diagonal density matrix.
+
+        Note: This is not the density matrix of the system, but is a convenient
+            way of computing expectation values of observables.
+
+        Returns:
+            qutip.Qobj: States probability distribution as a density matrix.
+        """
+        return self.get_state(-1)
+
+    def expect(self, obs_list: Sequence[Union[qutip.Qobj, ArrayLike]]
+               ) -> list[Union[float, complex, ArrayLike]]:
+        """Calculates the expectation value of a list of observables.
+
+        Args:
+            obs_list (array-like of qutip.Qobj or array-like of numpy.ndarray):
+                A list of observables whose expectation value will be
+                calculated. If necessary, each member will be transformed into
+                a qutip.Qobj instance.
+
+        Note: This only works for diagonal observables, since results have been
+            projected onto the Z basis.
+
+        Returns:
+            list: the list of expectation values of each operator.
+        """
+        for obs in obs_list:
+            if not isdiagonal(obs):
+                raise ValueError(f"Observable {obs} is non-diagonal.")
+
+        return super().expect(obs_list)
+
+    def _calc_weights(self, t: int) -> ArrayLike:
+        N = self._size
+        bitstrings = [np.binary_repr(k, N) for k in range(2**N)]
+        return [self._results[t][b] for b in bitstrings]
+
+    def _standard_dev(self, op: qutip.Qobj) -> ArrayLike:
+        """Returns the square root of the variance of operator op."""
+        density_mats = [self.get_state(t) for t in range(len(self._results))]
+        return cast(ArrayLike,
+                    np.sqrt(qutip.variance(op, density_mats) / self.N_measures)
+                    )
+
+    def _get_error_bars(self, op: qutip.Qobj) -> Tuple[ArrayLike, ArrayLike]:
+        moy = self.expect([op])[0]
+        st = self._standard_dev(op)
+        return moy, st
+
+    def plot(self, op: qutip.Qobj, fmt: str = '.',
+             label: str = '', error_bars: bool = True) -> None:
+        """Plots the expectation value of a given operator op.
+
+        Note: The observable must be diagonal.
+
+        Args:
+            op (qutip.Qobj): Operator whose expectation value is wanted.
+            error_bars (bool): Choose to display error bars.
+            fmt (str): Curve plot format.
+            label (str): y-Axis label.
+        """
+        if not isdiagonal(op):
+            raise ValueError(f"Observable {op} is non-diagonal.")
```

Comment:
Isn't this checked in `expect`?

Reply:
No, you can still have non-diagonal operators in the `CleanResults` case as no projective measurement has been done.

Reply:
That's right, this is `CleanResults`, my bad.